### DISCUSSION TITLE
bitbuffer decoding

### DISCUSSION THREAD

**Discussion by habub** [merbanan/rtl_433#2216] [General]
Hello,
Sorry if these are basic. Just warming up...

1) Bitbuffer
- How are the sync and gap words ("y=", "g=") reresented in the bitbuffer?
- What is the row delimiter in a bitbuffer?
- When/how does a bitbuffer end? What is the bitbuffer end delimiter?

2) Decoders
- How does the program decide which of the registered decoders (-R) to use for a signal?
- Can mutiple decoders be used on the same input signal? considering the signal can match more than 1 decoder?

Thanks!

**Comment by zuckschwerdt** [comment]
A sync or gap will break into a new row and sync also sets a sync counter (`syncs_before_row[i]`) on the row.
Rows have a given length (`bits_per_row[i]`). The Bitbuffers have a given number of rows (`num_rows`). S.a.
https://github.com/merbanan/rtl_433/blob/5f76eae6f6bf9c696b859296301bbbe234de77fe/include/bitbuffer.h#L32-L39

All decoders with matching "timings" are run. This means deciding if FSK / OOK is present and then trying if the timings can extract bits.

**Reply by habub** [reply]
Thanks!
I'm not sure I understand how to identify sync vs gap, and the reset signals.
When trying to decode [this signal](https://triq.org/pdv/#AAB03A07010778007C0190010011F83864B3B48191919191919191A291B3B39193919191B19193B39191B39191B39191B3B19193B3B391B193B3B1919555+AAB03907010778007C0190010011F83864B3B491919191919191A291B3B39191919191919391B1919191919391B1919191919391B19191919193919055+AAB05B07010778007C0190010011F83864B3B491919191919191A291B3B3919193B3B391B1919191919191939191B39191B193919191B193B391B1919193B193B3B19391B1939191B391B193B19391B1919193B191919193B191919191C655) assuming ook_pwm, I get:
{178} FE / 9F 6D B3 29 FF / 9F EF DF BF FF / 9C BF ED EA E9 BA B7 7B C (without setting gap)
Should this transmission be 1 bitbuffer or more? Does each "/" represent a new row?

**Reply by zuckschwerdt** [reply]
That's not PWM, there is data in the gaps. It's [PCM](https://triq.org/rtl_433/PULSE_FORMATS.html#nrz-l-%E2%80%94-non-return-to-zero-level) (125 µs) with [MC](https://triq.org/rtl_433/PULSE_FORMATS.html#mc-%E2%80%94-manchester-code), and there is a (de-)synch of 6 bit `111000` (which is not valid MC and thus a de-synch). Looking at that it's three or four transmissions or rows.

Use PCM with 125 short/long. Then put it [in a BitBench](https://triq.net/bitbench#c=FFFEAAAB8B32956A65656566A665A66B&c=55571665554B554B554B554A&c=0002AAAE2CCA9996AAA9595A55A65AA699A5A565A696A9AA9AAB&f=2b%20&a=Preamble&m=555c&cw=4) and break in the preamble, use a match (e.g. 55c) to align.
But you see the code is broken, invalid MC (there must not be an `00` or `11` parts), bad reception.

**Reply by habub** [reply]
Okay, wow....
I'll give it a try

**Reply by habub** [reply]
Hi,
Rgarding your reply:
> A sync or gap will break into a new row and sync also sets a sync counter (`syncs_before_row[i]`) on the row. Rows have a given length (`bits_per_row[i]`). The Bitbuffers have a given number of rows (`num_rows`). S.a.

Does this mean every decoder has a different bitbuffer structure? Since gaps and rows are unique for each decoder (y=, g=, r=). I thought bitbuffer is a more raw bitstream, prior to the deocding stage. Is this correct?

**Reply by zuckschwerdt** [reply]
The bitbuffer is a raw bitstream with some meta-data, e.g. where syncs and long gaps occured.

**Comment by habub** [comment]
One last question on this thread...

I decode the row values ( bitbuffer->bb[row] ) as a decimal number. How can I get (or convert) the Hex values of a row?
So it will appear like in the analyzers:
bitbuffer:: Number of rows: 2
[00] { 8} ff          : 11111111
[01] {26} 9e 37 ed c0 : 10011110 00110111 11101101 11

**Comment by zuckschwerdt** [comment]
Byte values are not decimal numbers, that's just one option to `printf` them (`"%d"`) and hex would be  `"%x"`.
To print debug values in a decoder you likely want
```
decoder_log_bitbuffer(decoder, 0, __func__, bitbuffer, "Rows Debug");
```

**Reply by habub** [reply]
I want to compare my data (in hex) to the bitbuffer data. Not print.
For example:

my_data = 0x 9E 37 ED C0
if ( my_data == bitbuffer->bb[i] ) ...

Which type should I declare my_data?

**Reply by merbanan** [reply]
This is a C-language question and this is not really the place for that. But if you want to compare you can use the uint8_t type.

uint8_t my_data = { 0x9E, 0x37 ....}

Then compare against bitbuffer->bb[0][0] which also should be uint8_t. But we have other tools for handling this. I suggest you read through a couple of decoders to see what is available.

**Reply by habub** [reply]
Thanks. You're correct, I'll brush up on my programming skills. It's been a long time...

**Reply by sp372** [reply]
interesting
